### PR TITLE
Server wont add or notify disconnected players anymore #6665

### LINKED ISSUE #6665: GhostRole System need to handle disconnected players
from looking through the code it looks like it doesn't have any way of handling, players that queue up then disconnect

### PR DESCRIPTION
**Pull request by MaxIsJoe**
fixes #6665

~~Need @Bod9001 to confirm that it solves the issue 100% incase I missed another queue function or a more important area that needed this fix.~~

Edit : we talked, this seems like it does the trick by making sure all disconnected players are cleared from the list before doing anything + the server wont waste time sending stuff to offline players.

### CODE REVIEW

**Comment by Bod9001** [comment]
looks good

**Review by Bod9001 — APPROVED** [review]

**Review by corp-0 — APPROVED** [review]

File: UnityProject/Assets/Scripts/Messages/Server/GhostRoles/GhostRoleUpdateMessage.cs [hunk 1]
```diff
@@ -39,6 +40,7 @@ public static NetMessage SendToDead(uint key)
 
 				foreach (ConnectedPlayer player in PlayerList.Instance.InGamePlayers)
 				{
+					if(PlayerList.Instance.loggedOff.Contains(player)) continue;
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Messages/Server/GhostRoles/GhostRoleUpdateMessage.cs:43]: This is unnecessary `PlayerList.Instance.InGamePlayers` is only loggedIn players.

`public List<ConnectedPlayer> InGamePlayers => loggedIn.FindAll(player => player.Script != null);`

File: UnityProject/Assets/Scripts/Systems/GhostRoles/GhostRoleManager.cs [hunk 2]
```diff
@@ -282,10 +285,23 @@ public void ServerRemoveWaitingPlayer(uint key, ConnectedPlayer player)
...
+		{
+			foreach (var ghostRoleQueue in serverAvailableRoles.Values)
+			{
+				foreach (var playerQueued in ghostRoleQueue.WaitingPlayers)
+				{
+					if(PlayerList.Instance.loggedOff.Contains(playerQueued) == false) continue;
+					ghostRoleQueue.WaitingPlayers.Remove(playerQueued);
+				}
+			}
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Systems/GhostRoles/GhostRoleManager.cs:303]: This will error as you are removing a item during iteration

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Systems/GhostRoles/GhostRoleManager.cs:303]: wait, when is it being iterated upon?

File: UnityProject/Assets/Scripts/Messages/Server/GhostRoles/GhostRoleUpdateMessage.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Messages/Server/GhostRoles/GhostRoleUpdateMessage.cs:43]: I'll add a comment to that variable so we can know that in the future

File: UnityProject/Assets/Scripts/Systems/GhostRoles/GhostRoleManager.cs (hunk 2 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Systems/GhostRoles/GhostRoleManager.cs:303]: `foreach (var playerQueued in ghostRoleQueue.WaitingPlayers)`

**Pull request merged**